envs.py

```diff
@@ -140,40 +143,49 @@ def step_async(self, actions):
     def step_wait(self):
         obs, reward, done, info = self.venv.step_wait()
         obs = torch.from_numpy(obs).float().to(self.device)
-        reward = torch.from_numpy(np.expand_dims(np.stack(reward), 1)).float()
+        reward = torch.from_numpy(reward).unsqueeze(dim=1).float()
         return obs, reward, done, info
 
 
 # Derived from
 # https://github.com/openai/baselines/blob/master/baselines/common/vec_env/vec_frame_stack.py
 class VecPyTorchFrameStack(VecEnvWrapper):
-    def __init__(self, venv, nstack, device):
+    def __init__(self, venv, nstack, device=None):
         self.venv = venv
         self.nstack = nstack
+
         wos = venv.observation_space  # wrapped ob space
-        self.shape_dim0 = wos.low.shape[0]
+        self.shape_dim0 = wos.shape[0]
+
         low = np.repeat(wos.low, self.nstack, axis=0)
         high = np.repeat(wos.high, self.nstack, axis=0)
-        self.stackedobs = np.zeros((venv.num_envs,) + low.shape)
-        self.stackedobs = torch.from_numpy(self.stackedobs).float()
-        self.stackedobs = self.stackedobs.to(device)
-        observation_space = gym.spaces.Box(low=low, high=high, dtype=venv.observation_space.dtype)
+
+        if device is None:
+            device = torch.device('cpu')
+        self.stacked_obs = torch.zeros((venv.num_envs,) + low.shape).to(device)
+
+        observation_space = gym.spaces.Box(
+            low=low, high=high, dtype=venv.observation_space.dtype)
         VecEnvWrapper.__init__(self, venv, observation_space=observation_space)
 
     def step_wait(self):
         obs, rews, news, infos = self.venv.step_wait()
-        self.stackedobs[:, :-self.shape_dim0] = self.stackedobs[:, self.shape_dim0:]
+        self.stacked_obs[:, :-self.shape_dim0] = \
+            self.stacked_obs[:, self.shape_dim0:]
         for (i, new) in enumerate(news):
             if new:
-                self.stackedobs[i] = 0
-        self.stackedobs[:, -self.shape_dim0:] = obs
-        return self.stackedobs, rews, news, infos
+                self.stacked_obs[i] = 0
+        self.stacked_obs[:, -self.shape_dim0:] = obs
+        return self.stacked_obs, rews, news, infos
 
     def reset(self):
         obs = self.venv.reset()
-        self.stackedobs.fill_(0)
-        self.stackedobs[:, -self.shape_dim0:] = obs
-        return self.stackedobs
+        self.stacked_obs.zero_()
+        self.stacked_obs[:, -self.shape_dim0:] = obs
+        return self.stacked_obs
 
     def close(self):
         self.venv.close()
+
+    def to(self, device):
```

Comment:
I'm not sure regarding this function. Since it diverge from the wrapper interface. What was the main reason to add it?

Reply:
For reasons of consistency, since all PyTorch objects that receive a `device` argument and have tensor variables have such a function. But one could of course just call this on the `stacked_obs` object itself. I removed it!